src/android/CHIPTool/app/src/main/java/com/google/chip/chiptool/clusterclient/SensorClientFragment.kt

```diff
@@ -0,0 +1,207 @@
+
+private typealias ReadCallback = ChipClusters.IntegerAttributeCallback
+
+class SensorClientFragment : Fragment() {
+  private val scope = CoroutineScope(Dispatchers.Main + Job())
+
+  // Timer to send periodic sensor read requests
+  private var sensorWatchTimer: Timer? = null
+
+  // History of sensor values
+  private val sensorData = LineGraphSeries<DataPoint>()
+
+  override fun onCreateView(
+      inflater: LayoutInflater,
+      container: ViewGroup?,
+      savedInstanceState: Bundle?
+  ): View {
+    return inflater.inflate(R.layout.sensor_client_fragment, container, false).apply {
+      clusterNameSpinner.adapter = makeClusterNamesAdapter()
+      clusterNameSpinner.onItemSelectedListener = object : AdapterView.OnItemSelectedListener {
+        override fun onNothingSelected(parent: AdapterView<*>?) = Unit
+        override fun onItemSelected(parent: AdapterView<*>?, view: View?, position: Int, id: Long) {
+          resetSensorGraph() // reset the graph on cluster change
+        }
+      }
+      readSensorBtn.setOnClickListener { scope.launch { readSensorButtonClick() } }
+      watchSensorBtn.setOnCheckedChangeListener { _, isChecked ->
+        if (isChecked) {
+          watchSensorButtonChecked()
+        } else {
+          watchSensorButtonUnchecked()
+        }
+      }
+      sensorGraph.addSeries(sensorData)
+      sensorGraph.gridLabelRenderer.padding = 30
+      sensorGraph.gridLabelRenderer.numHorizontalLabels = 4
+      sensorGraph.gridLabelRenderer.labelFormatter = object : LabelFormatter {
+        override fun setViewport(viewport: Viewport?) = Unit
+        override fun formatLabel(value: Double, isValueX: Boolean): String {
+          if (!isValueX)
+            return value.toString()
+          return SimpleDateFormat("H:m:s").format(Date(value.toLong())).toString()
+        }
+      }
+    }
+  }
+
+  override fun onStart() {
+    super.onStart()
+    deviceIdEd.setText(DeviceIdUtil.getLastDeviceId(requireContext()).toString())
+  }
+
+  override fun onStop() {
+    super.onStop()
+    scope.cancel()
+    resetSensorGraph() // reset the graph on fragment exit
+  }
+
+  private fun resetSensorGraph() {
+    watchSensorBtn.isChecked = false
+    sensorGraph.visibility = View.INVISIBLE
+    sensorData.resetData(emptyArray())
+  }
+
+  private fun makeClusterNamesAdapter(): ArrayAdapter<String> {
+    return ArrayAdapter(
+        requireContext(),
+        android.R.layout.simple_spinner_dropdown_item,
+        CLUSTERS.keys.toList()
+    ).apply {
+      setDropDownViewResource(android.R.layout.simple_spinner_dropdown_item)
+    }
+  }
+
+  private suspend fun readSensorButtonClick() {
+    try {
+      readSensorCluster(clusterNameSpinner.selectedItem.toString(), false)
+    } catch (ex: Exception) {
+      showMessage(R.string.sensor_client_read_error_text, ex.toString())
+    }
+  }
+
+  private fun watchSensorButtonChecked() {
+    sensorWatchTimer = Timer("SensorWatchTimer", false).apply {
```

Comment:
Since we're already using coroutines, can we use them to replace the Timer?
```
sensorWatchJob = scope.launch {
  while (isActive) {
    delay(REFRESH_PERIOD_MS)
    try { readSensorCluster(...) } catch (ex: Exception) { ... }
  }
}
```
You can `cancel()` the Job in `watchSensorButtonUnchecked()`.

LGTM for this PR otherwise

Reply:
Good point. I first wrote this fragment without coroutines and later switched to them, but I didn't update the timer. Resolved.